### PR TITLE
Add Invoice tooling

### PR DESCRIPTION
**Pull request by pixelfact**

### CODE REVIEW

File: shared/features/invoice/tokens/invoice-tokens.ts [hunk 1]
```diff
@@ -0,0 +1,65 @@
+export const InvoiceTokens = {
+  header: {
+    sampleTitle: "Your invoice number",
+    receiptTitle: "Receipt NO:",
+    invoiceTitle: "Invoice NO:",
+  },
+  invoiceInfos: {
+    wagmiName: "Wagmi",
+    wagmiAddress: "54 Rue Du Faubourg Montmartre, Paris, France, 75009",
```
Comment (haydencleary) [inline shared/features/invoice/tokens/invoice-tokens.ts:9]: Check with Camille but the address has changed https://onlydust.slack.com/archives/C031KBA3TJR/p1731405785428079

File: shared/features/invoice/tokens/invoice-tokens.ts [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+    siren: "SIREN",
+    vatNumber: "VAT Number",
+    issueDate: "Issue Date",
+    dueDate: "Due Date",
+    destinationAccounts: "Destination Accounts",
+    accountNumber: "Account number",
+    bic: "BIC",
+  },
+  rewardSummary: {
+    title: "Rewards granted on: app.onlydust.com",
```
Comment (haydencleary) [inline shared/features/invoice/tokens/invoice-tokens.ts:24]:
```suggestion
    title: "Rewards granted on: onlydust.com",
```
Seeing as the user can reward on admin.onlydust.com too ?

File: shared/features/invoice/hooks/use-invoice-upload/use-invoice-upload.tsx [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+      },
+    });
+
+  function handleSendInvoice({ fileBlob, isManualUpload = false, fileName }: HandleSendInvoiceProps) {
+    if (isManualUpload && fileName) {
+      const params = new URLSearchParams();
+      params.append("fileName", fileName);
+      setQueryParams(params);
+      capture("invoice_submitted", { type: "manuel" });
+    }
```
Comment (haydencleary) [inline shared/features/invoice/hooks/use-invoice-upload/use-invoice-upload.tsx:44]: Shouldn't there be a return here ?

File: shared/features/invoice/hooks/use-invoice-upload/use-invoice-upload.tsx [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+    if (fileBlob) {
+      uploadInvoice(fileBlob);
+      capture("invoice_submitted", { type: "auto-generated" });
+    } else {
```
Comment (haydencleary) [inline shared/features/invoice/hooks/use-invoice-upload/use-invoice-upload.tsx:48]: Maybe flatten these conditions with early returns ?

File: shared/features/invoice/hooks/use-invoice-preview/use-invoice-preview.tsx [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+        billingProfileId,
+        isSample,
+        impersonationHeaders: getHeaders(),
+      });
+      if (data?.blob) {
+        setFileBlob(data.blob);
+        setFileUrl(window.URL.createObjectURL(data.blob));
+        setInvoiceId(data.invoiceId ?? "");
+      } else {
+        setIsError(true);
```
Comment (haydencleary) [inline shared/features/invoice/hooks/use-invoice-preview/use-invoice-preview.tsx:43]: Perhaps throw to fall into the catch ?

File: shared/features/invoice/hooks/use-invoice-preview/use-invoice-preview.tsx [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+      }
+    } catch (error) {
+      setIsError(true);
```
Comment (haydencleary) [inline shared/features/invoice/hooks/use-invoice-preview/use-invoice-preview.tsx:46]: Set loading to false ?

File: shared/features/invoice/hooks/use-invoice-preview/use-invoice-preview.tsx [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+
+  useEffect(() => {
+    if (rewardIds && billingProfileId && !fetched.current) {
+      handleInvoiceCreation();
+    }
+  }, []);
+
+  async function handleInvoiceCreation() {
+    fetched.current = true;
+    setIsLoading(true);
```
Comment (haydencleary) [inline shared/features/invoice/hooks/use-invoice-preview/use-invoice-preview.tsx:28]: Set error to false ?

File: shared/features/invoice/handlers/fetch-invoice-preview-data.ts [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+      pathParams: {
+        billingProfileId,
+      },
+      queryParams: {
+        rewardIds: [rewardIds],
+      },
+      impersonationHeaders,
+    })
+    .request()
+    .then(res => res)
```
Comment (haydencleary) [inline shared/features/invoice/handlers/fetch-invoice-preview-data.ts:28]: Why ?

File: shared/features/invoice/_translations/invoice.en.json [hunk 9]
```diff
@@ -0,0 +1,18 @@
+{
+  "invoiceSubmission": {
+    "sidePanel": {
+      "title": "Send invoice"
+    },
+    "submitButton": "Add invoice",
+    "toaster": {
+      "success": "Invoice successfully sent",
+      "error": "We were unable to send the invoice, please try again later",
+      "emptyFile": "We Were unable to find the file, please try again later"
```
Comment (haydencleary) [inline shared/features/invoice/_translations/invoice.en.json:10]:
```suggestion
      "emptyFile": "We were unable to find the file, please try again later"
```

File: core/infrastructure/marketplace-api-client-adapter/adapters/billing-profile-client-adapter.ts [hunk 10]
```diff
@@ -54,4 +65,81 @@ export class BillingProfileClientAdapter implements BillingProfileStoragePort {
...
+    const request = async () => {
+      const data = await this.client.request<GetBillingProfileInvoicePreviewByIdResponse>({
+        path,
+        method,
+        tag,
+        pathParams,
+        queryParams,
+        headers: {
+          accept: "application/pdf",
+          ...(impersonationHeaders ? { "X-Impersonation-Claims": impersonationHeaders } : {}),
```
Comment (haydencleary) [inline core/infrastructure/marketplace-api-client-adapter/adapters/billing-profile-client-adapter.ts:86]: Impersonation is already handled inside `http-client`

File: app/api/invoice/route.tsx [hunk 11]
```diff
@@ -0,0 +1,126 @@
...
+  try {
+    invoicePreviewData = await fetchInvoicePreviewData({ token, rewardIds, billingProfileId, impersonationHeaders });
+  } catch (e) {
+    return new NextResponse("Failed Dependency : Invoice Preview ", { status: 424 });
+  }
+
+  /* ------
+  Determine the billing profile type
+  ------ */
+  const isUserIndividual = invoicePreviewData?.billingProfileType === BillingProfileType.Individual;
```
Comment (haydencleary) [inline app/api/invoice/route.tsx:49]: Could add this condition to the model ?

File: shared/features/invoice/handlers/fetch-font-info.ts [hunk 12]
```diff
@@ -0,0 +1,125 @@
...
+    if (fontFamilyMatch && srcMatch) {
+      return {
+        family: fontFamilyMatch[1],
+        fonts: [{ src: srcMatch[1], format: "truetype" }],
+      };
+    }
+
+    throw new Error("Font information not found");
+  } catch (error) {
+    console.error("Failed to fetch font info:", error);
```
Comment (haydencleary) [inline shared/features/invoice/handlers/fetch-font-info.ts:44]: Do we need logs here ?

File: shared/features/invoice/handlers/fetch-font-info.ts [hunk 13]
```diff
@@ -0,0 +1,125 @@
...
+  "prt", // Portugal
+  "rou", // Romania
+  "swe", // Sweden
+  "usa", // United States
+  "ury", // Uruguay
+];
+
+export async function detectLanguageAndGetFontDynamically(text: string): Promise<FontInfo> {
+  const languageCode = franc(text);
+  console.info("Detected language :", { languageCode, text });
```
Comment (haydencleary) [inline shared/features/invoice/handlers/fetch-font-info.ts:114]: Need logs ?

**Review by haydencleary — CHANGES_REQUESTED** [review]

File: core/infrastructure/marketplace-api-client-adapter/adapters/billing-profile-client-adapter.ts (hunk 10 above)
Comment (pixelfact) [reply core/infrastructure/marketplace-api-client-adapter/adapters/billing-profile-client-adapter.ts:86]: right!

File: shared/features/invoice/handlers/fetch-font-info.ts (hunk 13 above)
Comment (pixelfact) [reply shared/features/invoice/handlers/fetch-font-info.ts:114]: Yes if we have a font error on server we need to find what (exotic) language it is

File: shared/features/invoice/handlers/fetch-font-info.ts (hunk 12 above)
Comment (pixelfact) [reply shared/features/invoice/handlers/fetch-font-info.ts:44]: yes to easily find the error context on server

**Review by haydencleary — APPROVED** [review]

**Pull request merged**